packages/core/src/utils.js

```diff
@@ -1265,3 +1265,115 @@ export function schemaRequiresTrueValue(schema) {
 
   return false;
 }
+
+export function getRegistry(props) {
+  const { fields, widgets } = getDefaultRegistry();
+  return {
+    fields: { ...fields, ...props.fields },
+    widgets: { ...widgets, ...props.widgets },
+    ArrayFieldTemplate: props.ArrayFieldTemplate,
+    ObjectFieldTemplate: props.ObjectFieldTemplate,
+    FieldTemplate: props.FieldTemplate,
+    definitions: props.schema.definitions || {},
+    rootSchema: props.schema,
+    formContext: props.formContext || {},
+  };
+}
+
+export function validate(
+  formData,
+  props,
+  schema = props.schema,
+  additionalMetaSchemas = props.additionalMetaSchemas,
+  customFormats = props.customFormats
+) {
+  const { validate, transformErrors } = props;
+  const { rootSchema } = getRegistry(props);
+  const resolvedSchema = retrieveSchema(schema, rootSchema, formData);
+  return validateFormData(
+    formData,
+    resolvedSchema,
+    validate,
+    transformErrors,
+    additionalMetaSchemas,
+    customFormats
+  );
+}
+
+export function getStateFromProps(props, inputFormData, state = {}) {
+  const edit = typeof inputFormData !== "undefined";
```

Comment:
We should figure out -- why does this omit several lines from the original `getStateFromProps` function?

Reply:
the lines omitted are mostly just variable declarations that I pick up directly from the prop object instead